Feat/lobby chat

doganbros opened the pull request:
Better moderation capabilities in locked down rooms: challenge/response in lobby #9714
Lobby chat support
Knocking participants list updates
Knocking participants conditonal checks to show message button
Handle lobby chat events
Allow messages in mod_muc_lobby_rooms.lua

jitsi-jenkins commented:
Hi, thanks for your contribution!
If you haven't already done so, could you please make sure you sign our CLA (https://jitsi.org/icla for individuals and https://jitsi.org/ccla for corporations)? We would unfortunately be unable to merge your patch unless we have that piece :(.

doganbros commented:
@gpatel-fr we resolved most of the issues you raised in the previous pull request. 

1. We changed all the occurrences of challenge-response to lobby chat. Also the moderator is anonymous when he is chatting with the knocking participant.
2. Also we refactored the `mod_muc_lobby_rooms.lua`'s filter_stanza to support both private messages and moderator messages.
3. The code block you deleted the other time is still necessary because we need to make the attendee aware about the moderator's absence in the UI, even though they get the presence unavailable stanza.
4. We refactored some of the action names as you suggested.

damencho commented:
Hey @doganbros we deployed it here https://lobby-chat.jitsi.net/ so we can preview the feature ... but there is something wrong, when you click the chat button in the notification or participants pane it does not work ... can you check it out, please.

mkhstar commented:
@damencho probably the `mod_muc_lobby_rooms.lua` module was not updated. Because the messages are filtered

damencho commented:
> @damencho probably the `mod_muc_lobby_rooms.lua` module was not updated. Because the messages are filtered

Good point I will check it tomorrow, I think I restarted prosody... But maybe I missed that, will see.

damencho commented:
> > @damencho probably the `mod_muc_lobby_rooms.lua` module was not updated. Because the messages are filtered
> 
> Good point I will check it tomorrow, I think I restarted prosody... But maybe I missed that, will see.

Yep, you were right, now it works.

damencho commented on react/features/lobby/actions.any.js:
```diff
@@ -241,3 +250,149 @@ export function hideLobbyScreen() {
     };
 }
+
+/** ......................................
```
Can you drop the dots :)

damencho commented on react/features/lobby/components/web/LobbyScreen.js:
```diff
@@ -63,19 +108,46 @@ class LobbyScreen extends AbstractLobbyScreen {
     }
 
+
+    /** ....................
```
Drop and these, thank you.

damencho reviewed (changes requested):
lib-jitsi-meet part is merged :) Congrats and thank you!

damencho commented:
An update: I'm still waiting for a confirmation from Product and Design on few things. Design is finalizing the mobile part.
There are few changes probably. Here is what I know.
1. In the notification move Chat button between Admit and Reject.
![Screenshot 2022-02-07 at 14 43 31](https://user-images.githubusercontent.com/3263098/152870070-58c0bdeb-3597-4c5c-a403-2b30de62e8e2.png)

2. Make lobby chat use a different color than private messages (to be confirmed with the color).
3. Create overflow menu for Chat and Reject in participants pane.
![Screenshot 2022-02-07 at 14 45 21](https://user-images.githubusercontent.com/3263098/152870136-bfbb95be-55ce-4df1-b6e5-97735d90726f.png)

4. Create the chat for the guests to look like a normal group chat.
![Screenshot 2022-02-07 at 14 46 33](https://user-images.githubusercontent.com/3263098/152870167-7051f830-8f67-47e5-bfde-1d6d0204bd7a.png)

I think that's all. And when the design comes back to me, I will post here the link to the designs.
While waiting for the final thing. I think 1. and 3. can be done and now. For 3, the reasoning is too many buttons and those are not used so often and are visually hiding the display name. And there is already a button like that in the participants pane for other operations.
If 4 is not trivial ... we may leave it for the second phase or we can do it after merging this PR.

Thank you again.

mkhstar commented on resources/prosody-plugins/mod_muc_lobby_rooms.lua:
```diff
@@ -158,6 +158,20 @@ function filter_stanza(stanza)
         elseif stanza.name == 'iq' and stanza:get_child('query', DISCO_INFO_NS) then
             -- allow disco info from the lobby component
             return stanza;
+        elseif stanza.name == 'message' then
```
Hi @damencho we currently allow private messages and also messages only  to a moderator from line 163 to 173. (Or should we implement it differently). I think we are left with checking messages from a moderator

damencho replied:
Let me check again, when I was looking at it I had the impression that lobby users can send messages to lobby users privately or via group messages.

damencho replied:
So you allow all messages between lobby participants and group messages to moderators. So I was assuming group chats between participants are possible ... 

Can you easily check and do it: "messages to and from moderators are only allowed"... ?
It shouldn't be possible to find other participants ids, but just in case the check ^ seems better, I think.

mkhstar replied:
That makes sense, we will check on that

umitdogan1 commented:
> An update: I'm still waiting for a confirmation from Product and Design on few things. Design is finalizing the mobile part. There are few changes probably. Here is what I know.
> 
> 1. In the notification move Chat button between Admit and Reject.
>    ![Screenshot 2022-02-07 at 14 43 31](https://user-images.githubusercontent.com/3263098/152870070-58c0bdeb-3597-4c5c-a403-2b30de62e8e2.png)
> 2. Make lobby chat use a different color than private messages (to be confirmed with the color).
> 3. Create overflow menu for Chat and Reject in participants pane.
>    ![Screenshot 2022-02-07 at 14 45 21](https://user-images.githubusercontent.com/3263098/152870136-bfbb95be-55ce-4df1-b6e5-97735d90726f.png)
> 4. Create the chat for the guests to look like a normal group chat.
>    ![Screenshot 2022-02-07 at 14 46 33](https://user-images.githubusercontent.com/3263098/152870167-7051f830-8f67-47e5-bfde-1d6d0204bd7a.png)
> 
> I think that's all. And when the design comes back to me, I will post here the link to the designs. While waiting for the final thing. I think 1. and 3. can be done and now. For 3, the reasoning is too many buttons and those are not used so often and are visually hiding the display name. And there is already a button like that in the participants pane for other operations. If 4 is not trivial ... we may leave it for the second phase or we can do it after merging this PR.
> 
> Thank you again.

@damencho   I see... Will start to work on 1 and 3. Thank you too.

gpatel-fr commented:
Hello

I have not experienced any crash or unconsistent behaviour on my test instance.
I only tested on web, not mobile. 

A minor nit is that when there is one moderator and several candidates, the moderator workflow is not optimal.
If the mod sends a message to several persons, (without waiting for a reply of each), ideally it could be possible
to just wait for the answers and grant the entry without having to click again to chat with each person.
Well, the world is not perfect either but one day hopefully it will be :-)

On the code I have noticed this small strangeness in chat/actions.any.js:

 *  @param {boolean} removeLobbyChatMessages - The participant to remove

boolean for a participant ID ?? There is something not quite right here. 
I think that the comment should be changed to something like 

->  should remove messages from chat (works only for accepted users)

 * circa line 145 chat/action.any.js:  onLobbyChatInitialized(lobbyChatInititializedInfo... 
 spelling problem.
     -> lobbyChatInitializedInfo

 * in lobby/actions.any.js the function name 'updateLobbyParticipantOnModeratorLeave' seems to say that participantId should be a moderator's ID. But that's not the case, as shown in the code: 

if (knocking && lobbyMessageRecipient && lobbyMessageRecipient.id === participantId)

this part shows clearly that this function can be called with a non moderator ID.
So IMO the function could be called updateLobbyParticipantOnLeave instead.

(remark)
Note that I have wondered why this is working only for accepted participants, the rejected ones can still see the chat dialog after being left in the wild. That's because the participant is booted out of the lobby before it can be called. 
I don't think that's a problem, while it's good that the chat is cleaned for accepted participants (and this works).
What could be considered is that maybe the chat could be cleaned also for the moderator when a user is accepted/rejected.

On the prosody code, there is also a small nit: your code is duplicating a similar part in the
presence handling, however in this part the result is tested:

 local lobby_room = lobby_muc_service.get_room_from_jid(lobby_room_jid);
 if not lobby_room then
      module:log('warn', 'No lobby room found %s', lobby_room_jid);
      return stanza;
 end

if it makes sense to test the get_room_from_jid function in the presence handling, maybe it would make sense to test it also at this point.

Last - I would like to apologize for my first review, on one minor vocabulary point I was overly harsh and my tone was not appropriate. I can only say I was tired at the time.

Anyway, cheers for the very impressive work !

damencho commented:
@gpatel-fr Thank you for your review and for looking deep into the PR. That is a great help!

umitdogan1 commented:
> Hello
> 
> I have not experienced any crash or unconsistent behaviour on my test instance. I only tested on web, not mobile.
> 
> A minor nit is that when there is one moderator and several candidates, the moderator workflow is not optimal. If the mod sends a message to several persons, (without waiting for a reply of each), ideally it could be possible to just wait for the answers and grant the entry without having to click again to chat with each person. Well, the world is not perfect either but one day hopefully it will be :-)
> 
> On the code I have noticed this small strangeness in chat/actions.any.js:
> 
> * @param {boolean} removeLobbyChatMessages - The participant to remove
> 
> boolean for a participant ID ?? There is something not quite right here. I think that the comment should be changed to something like
> 
> -> should remove messages from chat (works only for accepted users)
> 
> * circa line 145 chat/action.any.js:  onLobbyChatInitialized(lobbyChatInititializedInfo...
>   spelling problem.
>   -> lobbyChatInitializedInfo
> * in lobby/actions.any.js the function name 'updateLobbyParticipantOnModeratorLeave' seems to say that participantId should be a moderator's ID. But that's not the case, as shown in the code:
> 
> if (knocking && lobbyMessageRecipient && lobbyMessageRecipient.id === participantId)
> 
> this part shows clearly that this function can be called with a non moderator ID. So IMO the function could be called updateLobbyParticipantOnLeave instead.
> 
> (remark) Note that I have wondered why this is working only for accepted participants, the rejected ones can still see the chat dialog after being left in the wild. That's because the participant is booted out of the lobby before it can be called. I don't think that's a problem, while it's good that the chat is cleaned for accepted participants (and this works). What could be considered is that maybe the chat could be cleaned also for the moderator when a user is accepted/rejected.
> 
> On the prosody code, there is also a small nit: your code is duplicating a similar part in the presence handling, however in this part the result is tested:
> 
> local lobby_room = lobby_muc_service.get_room_from_jid(lobby_room_jid); if not lobby_room then module:log('warn', 'No lobby room found %s', lobby_room_jid); return stanza; end
> 
> if it makes sense to test the get_room_from_jid function in the presence handling, maybe it would make sense to test it also at this point.
> 
> Last - I would like to apologize for my first review, on one minor vocabulary point I was overly harsh and my tone was not appropriate. I can only say I was tired at the time.
> 
> Anyway, cheers for the very impressive work !

Thank you on be half of the team. Will check your comments...

umitdogan1 commented:
@damencho few quick notes about the updates:

1. We did not change the color of "Reject" and left as it is (which is blue).
![image](https://user-images.githubusercontent.com/81876691/153826040-16bcc6bc-f018-4c95-b9a1-2a5c8337fdb5.png)

2. We tried to implement the buttons and overflow to match with the current UI implementation. Tried to be consistent with the current UI as much as possible. We can do this small styling/arrangements later when UI design is finalized in your side.

 2.1. Overflow menu icon with three dots is blue on the current jitsi-meet but it was gray on your screenshot. We left as it is.

 ![image](https://user-images.githubusercontent.com/81876691/153826115-4f37b966-0d87-43a2-9aec-85dd0fa62f46.png)

2.2. Also, changing the order of Reject and Admit in the current jitsi-meet UI would be better. But again we did not change it.

![image](https://user-images.githubusercontent.com/81876691/153826145-c9947b54-2866-42b9-949a-eee167c70053.png)

umitdogan1 commented:
@damencho any update regarding to new UI design?

damencho commented:
Hum, I was waiting for some replay ... Will check it tomorrow and will let you know.

damencho commented:
I think all I mentioned before is the latest state.

So I have these designs:
desktop: https://www.figma.com/file/xE4oxRXxQGJ9u1A56b4Oid/%E2%9C%8A-Challenge%2Fresponse-in-lobby?node-id=0%3A1
mobile: https://www.figma.com/file/xE4oxRXxQGJ9u1A56b4Oid/%E2%9C%8A-Challenge%2Fresponse-in-lobby?node-id=2%3A6400

Yeah do not do anything outside of the feature you are doing to keep it simple.

umitdogan1 commented:
> I think all I mentioned before is the latest state.
> 
> So I have these designs: desktop: https://www.figma.com/file/xE4oxRXxQGJ9u1A56b4Oid/%E2%9C%8A-Challenge%2Fresponse-in-lobby?node-id=0%3A1 mobile: https://www.figma.com/file/xE4oxRXxQGJ9u1A56b4Oid/%E2%9C%8A-Challenge%2Fresponse-in-lobby?node-id=2%3A6400
> 
> Yeah do not do anything outside of the feature you are doing to keep it simple.

Sure. Will check the designs.

umitdogan1 commented:
@damencho I guess the mobile designs [here](https://www.figma.com/file/xE4oxRXxQGJ9u1A56b4Oid/%E2%9C%8A-Challenge%2Fresponse-in-lobby?node-id=2%3A6400) are for the mobile screens of the web app (not mobile app), right? So, where are the mobile app designs?

mihaitiberiu commented:
> @damencho I guess the mobile designs [here](https://www.figma.com/file/xE4oxRXxQGJ9u1A56b4Oid/%E2%9C%8A-Challenge%2Fresponse-in-lobby?node-id=2%3A6400) are for the mobile screens of the web app (not mobile app), right? So, where are the mobile app designs?

-these designs should apply for both mobile web and native apps
-please ignore the "call quality" banner at the top since it's not available on Jitsi

mkhstar commented:
Also the chat screen of the design is quite different from the designs given. 
For example the text inputs, the chat bubble etc.
Screen shots:

<img width="1104" alt="chat-message-screen" src="https://user-images.githubusercontent.com/26695391/155842087-679fc120-a552-400e-bd8e-847b6a0d5262.png">

damencho commented:
Of course don't implement anything that is not for this feature. I think I had summarized earlier what are the needed changes. About the chat area in lobby is that there is no point of the marking of the messages with different colors, both for mobile and web. In lobby the chat should look like a normal chat in the room. 

@mihaitiberiu can you also chime in.

umitdogan1 commented:
@damencho Is it possible to give some priority to our PR? It will be perfect if we can ship the feature this week.

damencho commented:
Jenkins test this please

mkhstar commented:
Linting problems are fixed now

damencho commented:
Jenkins test this please

damencho commented:
I will try to look at it today and tomorrow.

damencho commented:
Jenkins test this please

damencho commented:
Congrats! Thank you for your contribution.

umitdogan1 commented:
Thank you too.
